lib/datasets.go

```diff
@@ -1736,22 +1179,534 @@ type datasetImpl struct{}
+		if err != nil {
+			return nil, err
+		}
+		err = maybeWriteOutfile(p, res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	} else if p.Selector == "stats" {
+		statsParams := &StatsParams{
+			Dataset: res.Dataset,
+		}
+		sa, err := scope.Loader().Dataset().Stats(scope.Context(), statsParams)
+		if err != nil {
+			return nil, err
+		}
+		res.Bytes, err = json.Marshal(sa.Stats)
+		if err != nil {
+			return nil, err
+		}
+		err = maybeWriteOutfile(p, res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
+
+	var value interface{}
+	if p.Selector == "" {
+		// `qri get` without a selector loads only the dataset head
+		value = res.Dataset
+	} else {
+		// `qri get <selector>` loads only the applicable component / field
+		value, err = base.ApplyPath(res.Dataset, p.Selector)
+		if err != nil {
+			return nil, err
+		}
+	}
+	switch p.Format {
+	case "json":
+		// Pretty defaults to true for the dataset head, unless explicitly set in the config.
+		pretty := true
+		if p.FormatConfig != nil {
+			pvalue, ok := p.FormatConfig.Map()["pretty"].(bool)
+			if ok {
+				pretty = pvalue
+			}
+		}
+		if pretty {
+			res.Bytes, err = json.MarshalIndent(value, "", " ")
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			res.Bytes, err = json.Marshal(value)
+			if err != nil {
+				return nil, err
+			}
+		}
+	case "yaml", "":
+		res.Bytes, err = yaml.Marshal(value)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unknown format: \"%s\"", p.Format)
+	}
+	err = maybeWriteOutfile(p, res)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Save adds a history entry, updating a dataset
+func (datasetImpl) Save(scope scope, p *SaveParams) (*dataset.Dataset, error) {
+	res := &dataset.Dataset{}
+
+	var (
+		writeDest = scope.Filesystem().DefaultWriteFS() // filesystem dataset will be written to
+		pro       = scope.Repo().Profiles().Owner()     // user making the request. hard-coded to repo owner
+	)
+
+	if p.Private {
+		return nil, fmt.Errorf("option to make dataset private not yet implemented, refer to https://github.com/qri-io/qri/issues/291 for updates")
+	}
+
+	// If the dscache doesn't exist yet, it will only be created if the appropriate flag enables it.
+	if p.UseDscache {
+		c := scope.Dscache()
+		c.CreateNewEnabled = true
+	}
+
+	// start with dataset fields provided by params
+	ds := p.Dataset
+	if ds == nil {
+		ds = &dataset.Dataset{}
+	}
+	ds.Assign(&dataset.Dataset{
+		BodyPath: p.BodyPath,
+		Commit: &dataset.Commit{
+			Title:   p.Title,
+			Message: p.Message,
+		},
+	})
+
+	if len(p.FilePaths) > 0 {
+		// TODO (b5): handle this with a qfs.Filesystem
+		dsf, err := ReadDatasetFiles(p.FilePaths...)
+		if err != nil {
+			return nil, err
+		}
+		dsf.Assign(ds)
+		ds = dsf
+	}
+
+	if p.Ref == "" && ds.Name != "" {
+		p.Ref = fmt.Sprintf("me/%s", ds.Name)
+	}
+
+	resolver, err := scope.Loader().resolverForMode("local")
```

Comment:
Add `ResolverForMode` to scope, whose body calls the `Instance` function, and use the scope version here.